src/resolve/target.rs

```diff
@@ -0,0 +1,331 @@
+                        .note(format!("canonicalizing the path: {:?}", path))
+                        .error(e.to_string())
+                        .emit();
+                    Err(Error::Diagnostic)
+                }
+            }
+        };
+
+        let inner = match inner {
+            inner @ TargetInner::Implementation(_) => inner,
+            inner @ TargetInner::Remote(_) => inner,
+            TargetInner::LocalAbsolute(abs) => TargetInner::LocalAbsolute(canonicalize(abs)?),
+            TargetInner::LocalRelative(rel) => {
+                assert!(rel.is_relative(), "TargetInner::LocalRelative not relative before canonicalizing: {:?}", rel);
+                let relative_to_project_root = meta.project_root.join(rel);
+                let canonicalized = canonicalize(relative_to_project_root)?;
+                TargetInner::LocalRelative(canonicalized)
+            },
+        };
+        Ok(TargetCanonicalized {
+            inner,
+            meta,
+        })
+    }
+}
+
+impl <'a, 'd> TargetCanonicalized<'a, 'd> {
+    /// Test if the source is allowed to request the target document.
+    ///
+    /// Some origins are not allowed to read all documents or only after explicit clearance by the
+    /// invoking user. Even more restrictive, the target handler could terminate the request at a
+    /// later time. For example when requesting a remote document make a CORS check.
+    pub fn check_access(self) -> Result<TargetChecked<'a, 'd>> {
+        let TargetCanonicalized { inner, meta } = self;
+        match (meta.context.typ(), &inner) {
+            (ContextType::LocalRelative, TargetInner::Implementation(_))
+            | (ContextType::LocalRelative, TargetInner::LocalRelative(_))
+            | (ContextType::LocalRelative, TargetInner::Remote(_)) => (),
+
+            (ContextType::LocalAbsolute, TargetInner::Implementation(_)) => (),
+            (ContextType::LocalAbsolute, TargetInner::LocalRelative(_))
+            | (ContextType::LocalAbsolute, TargetInner::Remote(_)) => {
+                meta.diagnostics
+                    .error("permission denied")
+                    .with_error_section(meta.range, "trying to include this")
+                    .note(
+                        "local absolute path not allowed to access remote or local relative files",
+                    )
+                    .emit();
+                return Err(Error::Diagnostic)
+            },
+
+            // TODO: discuss proper remote rules
+            // check CORS
+            (ContextType::Remote, TargetInner::Remote(url)) if meta.context.url.domain() == url.domain() => (),
+            (ContextType::Remote, TargetInner::Remote(_)) => {
+                meta.diagnostics
+                    .error("permission denied")
+                    .with_error_section(meta.range, "trying to include this")
+                    .error("CORS request detected")
```

Comment:
`CORS` hints to me that the resolver is a user-agent properly respecting the [`CORS`](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) http headers, which is not the case. I'd rather stay away from the acronym until we have figured out what to support and how (but it could be used in a note).

```suggestion
                    .error("cross origin request detected")
                    .note("proper CORS is not yet supported")
```